fix: invalid `txs_results` returned for legacy ABCI responses

andynog opened the pull request:
close: #3002 

This PR fixes the issue reported above.

This is not a storage issue in particular, the results are still in storage after an upgrade, but not returned properly by the RPC endpoint. The fix is to make the `/block_results` endpoint in `v0.38` to return properly a legacy ABCI response created with `v0.37`. 

Once this fix is merged on `v0.38` and a patch release is cut, any node on `v0.38` (e.g. an archive node) that applies the patch release, should have the results returned properly by the RPC `/block_results` endpoint. 

---

#### PR checklist

- [X] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] ~~Updated relevant documentation (`docs/` or `spec/`) and code comments~~
- [X] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

melekes reviewed (commented):
👍

sergio-mena commented on state/errors.go:
```diff
@@ -171,3 +171,7 @@ func (e ErrCannotLoadState) Error() string {
 	return e.Err
 }
+
+var ErrFinalizeBlockResponseUnmarshalError = errors.New("problem retrieving the finalize block response")
+
+var ErrFinalizeBlockResponseCorruptedError = errors.New("cannot retrieve finalize block response, corrupted data or spec change")
```
Can we add these in line 164-165?

andynog replied:
I will wrap this so these errors will change and will have Error() and Wrap() so will keep them separate

sergio-mena commented on state/store.go:
```diff
@@ -651,14 +651,22 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
-		if rerr != nil {
-			cmtos.Exit(fmt.Sprintf(`LoadFinalizeBlockResponse: Data has been corrupted or its spec has
-					changed: %v\n`, err))
+		rErr := legacyResp.Unmarshal(buf)
+		if rErr != nil {
```
```suggestion
		if err := legacyResp.Unmarshal(buf); err != nil {
```

sergio-mena commented on state/store.go:
```diff
@@ -651,14 +651,22 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
+		if rErr != nil {
+			// only return an error, this method is only invoked through the `/block_results` not for state logic and
+			// some tests, so no need to exit cometbft if there's an error, just return it.
+			return nil, ErrFinalizeBlockResponseCorruptedError
```
💯

sergio-mena commented on state/store.go:
```diff
@@ -651,14 +651,22 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
 	resp := new(abci.FinalizeBlockResponse)
 	err = resp.Unmarshal(buf)
+	// if an error is returned, then it needs to be handled here
 	if err != nil {
+		return nil, ErrFinalizeBlockResponseUnmarshalError
+	}
```
So, the bug is that a `LegacyABCIResponses` struct _can_ be unmarshalled as `FinalizeBlockResponse` with no error (albeit with incomplete information), is that right?

Also, if seems that this logic wasn't really unit-tested (this is coming from famous #9468), otherwise we would have caught it. Could you please add a UT that ... ?
* fails without this PR
* passes with this PR
🙏

andynog replied:
OK, I will add a test

andynog replied:
I've implemented a test under `api/api_test.go`, this test tries to create a message at the v0.37 level using the same proto definitions and then convert it to a legacy ABCI response. I think this is a good way to test this use cases across proto versions.

I look into creating a test using the store methods SaveABCIResponseInfo but the way the code is structured in store a lot of things are internal and cannot be used in the test, we'd need to refactor things quite a bit. So focusing on the proto messages I think is a good way to test. I might look into testing something generating a message in v0.37 and using the byte buffer in a test, that would simulate some the storage too.

andynog commented on state/store.go:
```diff
@@ -651,14 +651,22 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
 	err = resp.Unmarshal(buf)
+	// if an error is returned, then it needs to be handled here
 	if err != nil {
+		return nil, ErrFinalizeBlockResponseUnmarshalError
```
I've refactored the errors a bit. I didn't implement a `Wrap()` because it would surface the internal storage error to the API response call which is usually not a best practice. I think it's better to log the error from the store and then just return additional context (height) as the other errors are doing

andynog commented:
I am working on some additional tests to ensure the logic holds and it makes sense

cason commented on .changelog/unreleased/bug-fixes/3002-invalid-txs-results.md:
```diff
@@ -0,0 +1,3 @@
+- `[rpc]` Fix an issue where a legacy ABCI response, created on `v0.37` or before, is not returned properly in `v0.38` and up
+- on the `/block_results` RPC endpoint.
```
```suggestion
on the `/block_results` RPC endpoint.
```

cason commented on api/api_test.go:
I'd say that this test belongs to the `state` package. In any case, I would not keep it in this package...

andynog replied:
moved inside state in a `compatibility_test` file

cason commented on rpc/core/blocks.go:
```diff
@@ -193,6 +193,7 @@ func (env *Environment) BlockResults(_ *rpctypes.Context, heightPtr *int64) (*ct
 		FinalizeBlockEvents:   results.Events,
 		ValidatorUpdates:      results.ValidatorUpdates,
 		ConsensusParamUpdates: results.ConsensusParamUpdates,
+		AppHash:               results.AppHash,
```
Is this field returned via RPC? Or at least was it before (e.g., v0.38.x)?

melekes replied:
What about `LastResultsHash`? Should we include it too?

cason reviewed (commented):
No reserve against this change.

I would move the added test unit elsewhere thought.

andynog commented:
Just a quick update on this. I'm still doing some additional testing. The one I mentioned above was giving different results from other tests I'm performing with an "enhanced" kvstore (returns more information in a block_results). It seems the original logic was OK at least going from a `v0.37` to `v0.38`. Will add more context here once I have more information.

sergio-mena commented:
👍 let us know when you're done with the extra tests and I'll take another look

andynog commented:
> 👍 let us know when you're done with the extra tests and I'll take another look

Hi @sergio-mena, I've added an extra test `Test_V1Beta2_ABCIResponses_To_FinalizeBlockResponse` which un-marshalls the binary file (persisted ABCI Responses) as a FinalizeBlockResponse and if you check the test it is returning an error, so the old logic (`err != nil` instead of the suggested one I changed to `resp.AppHash == nil`) should still be good. 

I think in my troubleshooting above I was not getting the error because the `kvstore` response doesn't include data for all the fields, but once I generated a `v0.37` response (the `.bin` file) including DeliverTx, BeginBlock events I got the error when un-marshalling the binary as a `FinalizeBlockResponse`.

melekes commented on state/store.go:
```diff
@@ -651,14 +651,23 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
 	err = resp.Unmarshal(buf)
+	// if an error is returned, then it needs to be handled here
 	if err != nil {
+		store.Logger.Debug("failed to unmarshall ABCI response: %s", err.Error())
```
Why not include err in `ErrABCIResponseResponseUnmarshalForHeight`?

melekes commented on state/store.go:
```diff
@@ -651,14 +651,23 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
+		if err := legacyResp.Unmarshal(buf); err != nil {
+			// only return an error, this method is only invoked through the `/block_results` not for state logic and
+			// some tests, so no need to exit cometbft if there's an error, just return it.
+			store.Logger.Debug("failed to unmarshall legacy ABCI response: %s", err.Error())
```
same comment

sergio-mena commented on state/store.go:
```diff
@@ -651,14 +651,23 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
+		if err := legacyResp.Unmarshal(buf); err != nil {
+			// only return an error, this method is only invoked through the `/block_results` not for state logic and
+			// some tests, so no need to exit cometbft if there's an error, just return it.
+			store.Logger.Debug("failed to unmarshall legacy ABCI response: %s", err.Error())
```
```suggestion
			store.Logger.Debug("failed to unmarshall FinalizeBlockResponse (also tried as legacy ABCI response): %s", err.Error())
```

sergio-mena replied:
I would also include the error you get in line 662. The reason is, if it was indeed _not_ a legacy response, but it failed to deserialise, the error you get when you later try to deserialise as legacy is not useful.

sergio-mena commented on state/store.go:
```diff
@@ -651,14 +651,23 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
+			// only return an error, this method is only invoked through the `/block_results` not for state logic and
+			// some tests, so no need to exit cometbft if there's an error, just return it.
+			store.Logger.Debug("failed to unmarshall legacy ABCI response: %s", err.Error())
+			return nil, ErrABCIResponseCorruptedOrSpecChangeForHeight{Height: height}
```
Also, if we don't break any API (not sure) it'd be good to include the error inside this error, so the caller can print it if needed

andynog replied:
also logging as `Error()` not `Debug()`

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -0,0 +1,225 @@
+
+// Compatibility test across different state proto versions
+
+// This test unmarshal a v1beta2.ABCIResponses as a statev1.LegacyABCIResponses
```
```suggestion
// This test unmarshals a v1beta2.ABCIResponses as a statev1.LegacyABCIResponses
```

sergio-mena replied:
Same in comments below

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -0,0 +1,225 @@
+	}
+
+	responseDeliverTx := abciv1beta2.ResponseDeliverTx{
+		Code:   abci.CodeTypeOK,
```
Should you populate the rest of the fields here? (`GasWanted`, `Codespace`, etc)

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -0,0 +1,225 @@
+	require.NotNil(t, legacyABCIResponse.DeliverTxs)
+	require.NotNil(t, legacyABCIResponse.BeginBlock)
+	require.NotNil(t, legacyABCIResponse.EndBlock)
+	require.Equal(t, 1, len(legacyABCIResponse.DeliverTxs))
+	require.Equal(t, 1, len(legacyABCIResponse.BeginBlock.Events))
+	require.Equal(t, 1, len(legacyABCIResponse.EndBlock.Events))
```
I think we should be more exhaustive checking the values of these

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -0,0 +1,225 @@
+	// This cannot work since they have different schemas, a wrong wireType error is generated
+	finalizeBlockResponse := new(abci.FinalizeBlockResponse)
+	err = finalizeBlockResponse.Unmarshal(data)
+	require.Error(t, err)
+	require.ErrorContains(t, err, "unexpected EOF")
```
I don't understand, by looking at the production changes below, I had understood that the bug is that this unmarshalling actually doesn't error, but leaves some fields unfilled...

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -0,0 +1,225 @@
+	legacyResponseWithNull := new(statev1beta3.LegacyABCIResponses)
+	err = legacyResponseWithNull.Unmarshal(data)
+	require.NoError(t, err)
+	require.NotNil(t, legacyResponseWithNull.DeliverTxs)
+	require.Nil(t, legacyResponseWithNull.EndBlock)
+	require.NotNil(t, legacyResponseWithNull.BeginBlock)
```
Shouldn't we do the same checks here as in `TestStateProtoV1Beta2ToV1` above?

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -0,0 +1,225 @@
+package state_test
```
What I am missing here is a test case that exercises the (production) changes you made to `LoadFinalizeBlockResponse`.
In other words, if I take this PR and only apply the testing changes (without the production changes) is any test failing?

andynog replied:
Fixed it.

mbreithecker commented:
Please also add the `mode` attribute to the legacy response. Otherwise it is impossible to tell the difference between an BeginBlock and EndBlock event:

`state/store.go`

```go
// responseFinalizeBlockFromLegacy is a convenience function that takes the old abci responses and morphs
// it to the finalize block response. Note that the app hash is missing
func responseFinalizeBlockFromLegacy(legacyResp *cmtstate.LegacyABCIResponses) *abci.ResponseFinalizeBlock {

	// Add BeginBlock attribute to BeginBlock events
	for idx, event := range legacyResp.BeginBlock.Events {
		legacyResp.BeginBlock.Events[idx].Attributes = append(event.Attributes, abci.EventAttribute{
			Key:   "mode",
			Value: "BeginBlock",
			Index: false,
		})
	}

	// Add EndBlock attribute to BeginBlock events
	for idx, event := range legacyResp.EndBlock.Events {
		legacyResp.EndBlock.Events[idx].Attributes = append(event.Attributes, abci.EventAttribute{
			Key:   "mode",
			Value: "EndBlock",
			Index: false,
		})
	}

	return &abci.ResponseFinalizeBlock{
		TxResults:             legacyResp.DeliverTxs,
		ValidatorUpdates:      legacyResp.EndBlock.ValidatorUpdates,
		ConsensusParamUpdates: legacyResp.EndBlock.ConsensusParamUpdates,
		Events:                append(legacyResp.BeginBlock.Events, legacyResp.EndBlock.Events...),
		// NOTE: AppHash is missing in the response but will
		// be caught and filled in consensus/replay.go
	}
}
```

johnletey commented:
Noble would love to see this released as well!

andynog commented:
Just an update, still working on this, need to add some additional testing to ensure these changes are good especially backporting to old releases. But hopefully should be ready to be merged sooner than later.

andynog commented:
Just an update. The fix is still on-going. We believe we have a solution but might be hard to test with live production data (archive nodes). Would anyone that benefits from this fix be able to test against real data before we cut an official release ? 

Once the fix is in, we could backport to v0.38.x branch (but we would not cut a release until we have some confirmation the fix works). Then you can test using a binary compiled from that branch and let us know.

jhernandezb commented:
hi @andynog we could give it a try in our testnet but we will need what @mbreithecker suggested in https://github.com/cometbft/cometbft/pull/3031#issuecomment-2152095271 to be able to identify being/end block events

andynog commented:
Hi @mbreithecker and @jhernandez,

I made some changes a few days ago, including the attributes logic suggested. Do you need to test this on a v0.38 level branch, or are you okay with cherry-picking the changes? Please let me know. Thanks for your help!

mmsqe commented:
Thanks @andynog, patch works for v0.38.x.

melekes reviewed (approved):
👍

SpicyLemon commented on state/store.go:
```diff
@@ -653,14 +653,22 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
+		if err := legacyResp.Unmarshal(buf); err != nil {
+			// only return an error, this method is only invoked through the `/block_results` not for state logic and
+			// some tests, so no need to exit cometbft if there's an error, just return it.
+			store.Logger.Debug("failed to unmarshall FinalizeBlockResponse (also tried as legacy ABCI response): %s", err.Error())
```
The log methods (e.g. `Debug`) take in a string message, then alternating key, value args. You should either wrap the two args in a `fmt.Sprintf` or else remove the `: %s` from the message and add an `"error"` key argument before `err.Error()`.


```suggestion
			store.Logger.Debug(fmt.Sprintf("failed to unmarshall FinalizeBlockResponse (also tried as legacy ABCI response): %v", err))
```

or 

```suggestion
			store.Logger.Debug("failed to unmarshall FinalizeBlockResponse (also tried as legacy ABCI response)", "error", err)
```

melekes commented:
@andynog can we merge this now since this patch was verified?

andynog commented:
> @andynog can we merge this now since this patch was verified?

Hi @melekes, I believe we can merge since we got a couple of confirmations that the patch works. One confirmation here and another [here](https://github.com/cometbft/cometbft/issues/3354#issuecomment-2201563686)

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
-	abciResponses := newV1Beta2ABCIResponses()
-	require.Equal(t, 1, len(abciResponses.DeliverTxs))
-	require.Equal(t, 1, len(abciResponses.BeginBlock.Events))
-	require.Equal(t, 1, len(abciResponses.EndBlock.Events))
```
Could you keep these three tests? This way we make sure that e.g. check in line 157 isn't passing because 0 == 0... in this case line 158 will panic

andynog replied:
good point, I've added the tests

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
+	require.Nil(t, finalizeBlockResponse.AppHash)
+
+	// Test for equality
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs), len(finalizeBlockResponse.TxResults))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Code, finalizeBlockResponse.TxResults[0].Code)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Data, finalizeBlockResponse.TxResults[0].Data)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Log, finalizeBlockResponse.TxResults[0].Log)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasWanted, finalizeBlockResponse.TxResults[0].GasWanted)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasUsed, finalizeBlockResponse.TxResults[0].GasUsed)
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events), len(finalizeBlockResponse.TxResults[0].Events))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Type, finalizeBlockResponse.TxResults[0].Events[0].Type)
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes), len(finalizeBlockResponse.TxResults[0].Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key, finalizeBlockResponse.TxResults[0].Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, finalizeBlockResponse.TxResults[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, finalizeBlockResponse.TxResults[0].Codespace)
```
Excellent! Just one question (not sure about this): if you call
```
require.Equal(v1beta2ABCIResponses.DeliverTxs, finalizeBlockResponse.TxResults)
```
Won't that check all the values and all the fields in the values?

sergio-mena replied:
Actually, that's what you're doing in line 211 below (with a different value for `v1beta2ABCIResponses`)

andynog replied:
I can't check them for equality straight away because they are based on different protos so the check fails (I tried that first), because even though structurally they might look the same their fields might come from a different version and the `Equals` fails

andynog replied:
```
=== RUN   TestSaveLegacyAndLoadFinalizeBlock
    compatibility_test.go:158: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:158
        	Error:      	Not equal: 
        	            	expected: []*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x140002e0380)})
        	            	actual  : []*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x140002e0400)})
        	Test:       	TestSaveLegacyAndLoadFinalizeBlock
--- FAIL: TestSaveLegacyAndLoadFinalizeBlock (0.00s)

Expected :[]*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x140002e0380)})
Actual   :[]*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x140002e0400)})
```

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
+	require.NotNil(t, legacyABCIResponses.BeginBlock)
+
+	// ensure for equality
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs), len(legacyABCIResponses.DeliverTxs))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Code, legacyABCIResponses.DeliverTxs[0].Code)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Data, legacyABCIResponses.DeliverTxs[0].Data)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Log, legacyABCIResponses.DeliverTxs[0].Log)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasWanted, legacyABCIResponses.DeliverTxs[0].GasWanted)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasUsed, legacyABCIResponses.DeliverTxs[0].GasUsed)
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events), len(legacyABCIResponses.DeliverTxs[0].Events))
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes), len(legacyABCIResponses.DeliverTxs[0].Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, legacyABCIResponses.DeliverTxs[0].Codespace)
```
Same comment as above: wouldn't `require.Equal(t, v1beta2ABCIResponses.DeliverTxs, legacyABCIResponses.DeliverTxs)` be checking all elements and all fields one by one?

andynog replied:
Same answer as above, but here's the error if you try to test for equality

```
=== RUN   TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse
    compatibility_test.go:239: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:239
        	Error:      	Not equal: 
        	            	expected: []*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000255b00)})
        	            	actual  : []*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x14000255b80)})
        	Test:       	TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse
--- FAIL: TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse (0.00s)

Expected :[]*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000255b00)})
Actual   :[]*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x14000255b80)})```

sergio-mena replied:
Thanks for giving it a try

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, legacyABCIResponses.DeliverTxs[0].Codespace)
+
+	require.Equal(t, len(v1beta2ABCIResponses.BeginBlock.Events), len(legacyABCIResponses.BeginBlock.Events))
```
Same here, with `require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events, legacyABCIResponses.BeginBlock.Events`

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, legacyABCIResponses.DeliverTxs[0].Codespace)
+
+	require.Equal(t, len(v1beta2ABCIResponses.BeginBlock.Events), len(legacyABCIResponses.BeginBlock.Events))
+	require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events[0].Type, legacyABCIResponses.BeginBlock.Events[0].Type)
+	require.Equal(t, len(v1beta2ABCIResponses.BeginBlock.Events[0].Attributes), len(legacyABCIResponses.BeginBlock.Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events[0].Attributes[0].Key, legacyABCIResponses.BeginBlock.Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events[0].Attributes[0].Value, legacyABCIResponses.BeginBlock.Events[0].Attributes[0].Value)
```
Same here, with `require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events, legacyABCIResponses.BeginBlock.Events`

andynog replied:
same as above

andynog replied:
```
=== RUN   TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse
    compatibility_test.go:252: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:252
        	Error:      	Not equal: 
        	            	expected: []v1beta2.Event([]v1beta2.Event{v1beta2.Event{Type:"begin_block_event", Attributes:[]v1beta2.EventAttribute{v1beta2.EventAttribute{Key:"key", Value:"value", Index:false}}}})
        	            	actual  : []v1.Event([]v1.Event{v1.Event{Type:"begin_block_event", Attributes:[]v1.EventAttribute{v1.EventAttribute{Key:"key", Value:"value", Index:false}}}})
        	Test:       	TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse
--- FAIL: TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse (0.00s)

Expected :[]v1beta2.Event([]v1beta2.Event{v1beta2.Event{Type:"begin_block_event", Attributes:[]v1beta2.EventAttribute{v1beta2.EventAttribute{Key:"key", Value:"value", Index:false}}}})
Actual   :[]v1.Event([]v1.Event{v1.Event{Type:"begin_block_event", Attributes:[]v1.EventAttribute{v1.EventAttribute{Key:"key", Value:"value", Index:false}}}})```

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
+	require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events[0].Attributes[0].Value, legacyABCIResponses.BeginBlock.Events[0].Attributes[0].Value)
+}
+
+// This test unmarshal a v1beta2.ABCIResponses as a v1beta3.ResponseFinalizeBlock
```
```suggestion
// This test unmarshals a v1beta2.ABCIResponses as a v1beta3.ResponseFinalizeBlock
```

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -204,15 +324,36 @@ func TestStateV1Beta2ABCIResponsesAsStateV1Beta3FinalizeBlockResponse(t *testing
 	require.NotNil(t, legacyABCIResponses.BeginBlock)
+
+	// ensure for equality
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs), len(legacyABCIResponses.DeliverTxs))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Code, legacyABCIResponses.DeliverTxs[0].Code)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Data, legacyABCIResponses.DeliverTxs[0].Data)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Log, legacyABCIResponses.DeliverTxs[0].Log)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasWanted, legacyABCIResponses.DeliverTxs[0].GasWanted)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasUsed, legacyABCIResponses.DeliverTxs[0].GasUsed)
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events), len(legacyABCIResponses.DeliverTxs[0].Events))
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes), len(legacyABCIResponses.DeliverTxs[0].Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, legacyABCIResponses.DeliverTxs[0].Codespace)
```
Same comment

andynog replied:
same as above

andynog replied:
```
=== RUN   TestStateV1Beta2ABCIResponsesAsStateV1Beta3LegacyABCIResponse
    compatibility_test.go:335: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:335
        	Error:      	Not equal: 
        	            	expected: []*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000316400)})
        	            	actual  : []*v1beta3.ExecTxResult([]*v1beta3.ExecTxResult{(*v1beta3.ExecTxResult)(0x14000316480)})
        	Test:       	TestStateV1Beta2ABCIResponsesAsStateV1Beta3LegacyABCIResponse
--- FAIL: TestStateV1Beta2ABCIResponsesAsStateV1Beta3LegacyABCIResponse (0.00s)

Expected :[]*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000316400)})
Actual   :[]*v1beta3.ExecTxResult([]*v1beta3.ExecTxResult{(*v1beta3.ExecTxResult)(0x14000316480)})```

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -204,15 +324,36 @@ func TestStateV1Beta2ABCIResponsesAsStateV1Beta3FinalizeBlockResponse(t *testing
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, legacyABCIResponses.DeliverTxs[0].Codespace)
+
+	require.Equal(t, len(v1beta2ABCIResponses.BeginBlock.Events), len(legacyABCIResponses.BeginBlock.Events))
+	require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events[0].Type, legacyABCIResponses.BeginBlock.Events[0].Type)
+	require.Equal(t, len(v1beta2ABCIResponses.BeginBlock.Events[0].Attributes), len(legacyABCIResponses.BeginBlock.Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events[0].Attributes[0].Key, legacyABCIResponses.BeginBlock.Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponses.BeginBlock.Events[0].Attributes[0].Value, legacyABCIResponses.BeginBlock.Events[0].Attributes[0].Value)
```
Same comment

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -226,6 +367,89 @@ func TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3FinalizeBlockResponse(t
 	require.Nil(t, legacyResponseWithNull.EndBlock)
 	require.NotNil(t, legacyResponseWithNull.BeginBlock)
+
+	require.Equal(t, len(v1beta2ABCIResponsesWithNull.BeginBlock.Events), len(legacyResponseWithNull.BeginBlock.Events))
+	require.Equal(t, v1beta2ABCIResponsesWithNull.BeginBlock.Events[0].Type, legacyResponseWithNull.BeginBlock.Events[0].Type)
+	require.Equal(t, len(v1beta2ABCIResponsesWithNull.BeginBlock.Events[0].Attributes), len(legacyResponseWithNull.BeginBlock.Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponsesWithNull.BeginBlock.Events[0].Attributes[0].Key, legacyResponseWithNull.BeginBlock.Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.BeginBlock.Events[0].Attributes[0].Value, legacyResponseWithNull.BeginBlock.Events[0].Attributes[0].Value)
```
Same

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -226,6 +367,89 @@ func TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3FinalizeBlockResponse(t
+	require.Equal(t, v1beta2ABCIResponsesWithNull.BeginBlock.Events[0].Attributes[0].Key, legacyResponseWithNull.BeginBlock.Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.BeginBlock.Events[0].Attributes[0].Value, legacyResponseWithNull.BeginBlock.Events[0].Attributes[0].Value)
+
+	require.Equal(t, len(v1beta2ABCIResponsesWithNull.DeliverTxs), len(legacyResponseWithNull.DeliverTxs))
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].Code, legacyResponseWithNull.DeliverTxs[0].Code)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].Data, legacyResponseWithNull.DeliverTxs[0].Data)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].Log, legacyResponseWithNull.DeliverTxs[0].Log)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].GasWanted, legacyResponseWithNull.DeliverTxs[0].GasWanted)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].GasUsed, legacyResponseWithNull.DeliverTxs[0].GasUsed)
+	require.Equal(t, len(v1beta2ABCIResponsesWithNull.DeliverTxs[0].Events), len(legacyResponseWithNull.DeliverTxs[0].Events))
+	require.Equal(t, len(v1beta2ABCIResponsesWithNull.DeliverTxs[0].Events[0].Attributes), len(legacyResponseWithNull.DeliverTxs[0].Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].Events[0].Attributes[0].Key, legacyResponseWithNull.DeliverTxs[0].Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].Events[0].Attributes[0].Value, legacyResponseWithNull.DeliverTxs[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponsesWithNull.DeliverTxs[0].Codespace, legacyResponseWithNull.DeliverTxs[0].Codespace)
```
Same

andynog replied:
```
=== RUN   TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3LegacyABCIResponse
    compatibility_test.go:380: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:380
        	Error:      	Not equal: 
        	            	expected: []*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000326600)})
        	            	actual  : []*v1beta3.ExecTxResult([]*v1beta3.ExecTxResult{(*v1beta3.ExecTxResult)(0x14000326680)})
        	Test:       	TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3LegacyABCIResponse
--- FAIL: TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3LegacyABCIResponse (0.00s)

Expected :[]*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000326600)})
Actual   :[]*v1beta3.ExecTxResult([]*v1beta3.ExecTxResult{(*v1beta3.ExecTxResult)(0x14000326680)})```

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -226,6 +367,89 @@ func TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3FinalizeBlockResponse(t
+	require.NotNil(t, finalizeBlockResponse.AppHash)
+
+	// Test for equality
+	require.Equal(t, len(v1beta3ResponseFinalizeBlock.TxResults), len(finalizeBlockResponse.TxResults))
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Code, finalizeBlockResponse.TxResults[0].Code)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Data, finalizeBlockResponse.TxResults[0].Data)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Log, finalizeBlockResponse.TxResults[0].Log)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].GasWanted, finalizeBlockResponse.TxResults[0].GasWanted)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].GasUsed, finalizeBlockResponse.TxResults[0].GasUsed)
+	require.Equal(t, len(v1beta3ResponseFinalizeBlock.TxResults[0].Events), len(finalizeBlockResponse.TxResults[0].Events))
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Events[0].Type, finalizeBlockResponse.TxResults[0].Events[0].Type)
+	require.Equal(t, len(v1beta3ResponseFinalizeBlock.TxResults[0].Events[0].Attributes), len(finalizeBlockResponse.TxResults[0].Events[0].Attributes))
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Events[0].Attributes[0].Key, finalizeBlockResponse.TxResults[0].Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Events[0].Attributes[0].Value, finalizeBlockResponse.TxResults[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Codespace, finalizeBlockResponse.TxResults[0].Codespace)
```
same

andynog replied:
```
=== RUN   TestStateV1Beta3ResponsesFinalizeBlockAsV1FinalizeBlockResponse
    compatibility_test.go:416: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:416
        	Error:      	Not equal: 
        	            	expected: []*v1beta3.ExecTxResult([]*v1beta3.ExecTxResult{(*v1beta3.ExecTxResult)(0x1400030ff00)})
        	            	actual  : []*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x14000316000)})
        	Test:       	TestStateV1Beta3ResponsesFinalizeBlockAsV1FinalizeBlockResponse
--- FAIL: TestStateV1Beta3ResponsesFinalizeBlockAsV1FinalizeBlockResponse (0.00s)

Expected :[]*v1beta3.ExecTxResult([]*v1beta3.ExecTxResult{(*v1beta3.ExecTxResult)(0x1400030ff00)})
Actual   :[]*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x14000316000)})
```

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -226,6 +367,89 @@ func TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3FinalizeBlockResponse(t
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Events[0].Attributes[0].Value, finalizeBlockResponse.TxResults[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.TxResults[0].Codespace, finalizeBlockResponse.TxResults[0].Codespace)
+
+	require.Equal(t, len(v1beta3ResponseFinalizeBlock.Events), len(finalizeBlockResponse.Events))
+	require.Equal(t, v1beta3ResponseFinalizeBlock.Events[0].Type, finalizeBlockResponse.Events[0].Type)
+	require.Equal(t, len(v1beta3ResponseFinalizeBlock.Events[0].Attributes), len(finalizeBlockResponse.Events[0].Attributes))
+	require.Equal(t, v1beta3ResponseFinalizeBlock.Events[0].Attributes[0].Key, finalizeBlockResponse.Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.Events[0].Attributes[0].Value, finalizeBlockResponse.Events[0].Attributes[0].Value)
```
same

andynog replied:
```
=== RUN   TestStateV1Beta3ResponsesFinalizeBlockAsV1FinalizeBlockResponse
    compatibility_test.go:429: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:429
        	Error:      	Not equal: 
        	            	expected: []v1beta2.Event([]v1beta2.Event{v1beta2.Event{Type:"tx_event", Attributes:[]v1beta2.EventAttribute{v1beta2.EventAttribute{Key:"key", Value:"value", Index:false}}}, v1beta2.Event{Type:"one_event", Attributes:[]v1beta2.EventAttribute{v1beta2.EventAttribute{Key:"key", Value:"value", Index:false}}}, v1beta2.Event{Type:"two_event", Attributes:[]v1beta2.EventAttribute{v1beta2.EventAttribute{Key:"key", Value:"value", Index:false}}}})
        	            	actual  : []v1.Event([]v1.Event{v1.Event{Type:"tx_event", Attributes:[]v1.EventAttribute{v1.EventAttribute{Key:"key", Value:"value", Index:false}}}, v1.Event{Type:"one_event", Attributes:[]v1.EventAttribute{v1.EventAttribute{Key:"key", Value:"value", Index:false}}}, v1.Event{Type:"two_event", Attributes:[]v1.EventAttribute{v1.EventAttribute{Key:"key", Value:"value", Index:false}}}})
        	Test:       	TestStateV1Beta3ResponsesFinalizeBlockAsV1FinalizeBlockResponse
--- FAIL: TestStateV1Beta3ResponsesFinalizeBlockAsV1FinalizeBlockResponse (0.00s)

Expected :[]v1beta2.Event([]v1beta2.Event{v1beta2.Event{Type:"tx_event", Attributes:[]v1beta2.EventAttribute{v1beta2.EventAttribute{Key:"key", Value:"value", Index:false}}}, v1beta2.Event{Type:"one_event", Attributes:[]v1beta2.EventAttribute{v1beta2.EventAttri ...

Actual   :[]v1.Event([]v1.Event{v1.Event{Type:"tx_event", Attributes:[]v1.EventAttribute{v1.EventAttribute{Key:"key", Value:"value", Index:false}}}, v1.Event{Type:"one_event", Attributes:[]v1.EventAttribute{v1.EventAttribute{Key:"key", Value:"value", Index:fal ...
```

sergio-mena commented on state/compatibility_test.go:
```diff
@@ -226,6 +367,89 @@ func TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3FinalizeBlockResponse(t
+	require.Equal(t, v1beta3ResponseFinalizeBlock.Events[0].Attributes[0].Key, finalizeBlockResponse.Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.Events[0].Attributes[0].Value, finalizeBlockResponse.Events[0].Attributes[0].Value)
+
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Block.MaxBytes, finalizeBlockResponse.ConsensusParamUpdates.Block.MaxBytes)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Block.MaxGas, finalizeBlockResponse.ConsensusParamUpdates.Block.MaxGas)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Evidence.MaxAgeNumBlocks, finalizeBlockResponse.ConsensusParamUpdates.Evidence.MaxAgeNumBlocks)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Evidence.MaxAgeDuration, finalizeBlockResponse.ConsensusParamUpdates.Evidence.MaxAgeDuration)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Evidence.MaxBytes, finalizeBlockResponse.ConsensusParamUpdates.Evidence.MaxBytes)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Validator.PubKeyTypes, finalizeBlockResponse.ConsensusParamUpdates.Validator.PubKeyTypes)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Version.App, finalizeBlockResponse.ConsensusParamUpdates.Version.App)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Synchrony.Precision, finalizeBlockResponse.ConsensusParamUpdates.Synchrony.Precision)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Synchrony.MessageDelay, finalizeBlockResponse.ConsensusParamUpdates.Synchrony.MessageDelay)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Feature.VoteExtensionsEnableHeight.Value, finalizeBlockResponse.ConsensusParamUpdates.Feature.VoteExtensionsEnableHeight.Value)
+	require.Equal(t, v1beta3ResponseFinalizeBlock.ConsensusParamUpdates.Feature.PbtsEnableHeight.Value, finalizeBlockResponse.ConsensusParamUpdates.Feature.PbtsEnableHeight.Value)
```
Same: should comparing `v1beta3ResponseFinalizeBlock.ConsensusParamUpdates` and `finalizeBlockResponse.ConsensusParamUpdates` do the same check as these?

andynog replied:
That can work. I'll add the test. It just works because `v1beta3ResponseFinalizeBlock.ConsensusParamUpdates` uses a `v1.ConsensusParams` the same used by `finalizeBlockResponse.ConsensusParamUpdates` but this might change in the future because of #3090. So I'll add the equality test but will leave the individual ones too.

sergio-mena reviewed (approved):
Thanks a lot for this!

andynog commented:
Hi @mbreithecker, @mmsqe, @johnletey, @jhernandezb and @SpicyLemon 

This fixed has been back ported to `v0.38.x` (#3435) and it will be available in the next release in the `v0.38` line. Thanks for all feedback and testing. Please let us know if you encounter any issues while using the fix.

Thanks again,

Andy.

mbreithecker commented:
Thank you very much!
